macaroon_service.go

```diff
@@ -0,0 +1,287 @@
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/macaroons"
+	"github.com/lightningnetwork/lnd/rpcperms"
+	"go.etcd.io/bbolt"
+	"google.golang.org/grpc"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+	"gopkg.in/macaroon-bakery.v2/bakery/checkers"
+)
+
+const (
+	// defaultDBName is the default macaroon db name.
+	defaultDBName = "macaroon.db"
+
+	// defaultDBTimeout is the default timeout to be used for the
+	// macaroon db connection.
+	defaultDBTimeout = 5 * time.Second
+)
+
+// MacaroonService handles the creatation and unlocking of a macaroon DB file.
+// It is also a wrapper for a macaroon.Service and uses this to create a
+// default macaroon for the caller if stateless mode has not been specified.
+type MacaroonService struct {
+	cfg *MacaroonServiceConfig
+
+	*macaroons.Service
+}
+
+// MacaroonServiceConfig holds configuration values used by the MacaroonService.
+type MacaroonServiceConfig struct {
+	// DBPath is the path to where the macaroon db file will be stored.
+	DBPath string
+
+	// DBFile is the name of the macaroon db.
+	DBFileName string
+
+	// DBTimeout is the maximum time we wait for the bbolt database to be
+	// opened.
+	DBTimeout time.Duration
+
+	// MacaroonLocation is the value used for a macaroons' "Location" field.
+	MacaroonLocation string
+
+	// MacaroonPath is the path to where macaroons should be stored.
+	MacaroonPath string
+
+	// StatelessInit should be set to true if no default macaroons should
+	// be created and stored on disk.
+	StatelessInit bool
+
+	// Checkers are used to add extra validation checks on macaroon.
+	Checkers []macaroons.Checker
+
+	// RequiredPerms defines all method paths and the permissions required
+	// when accessing those paths.
+	RequiredPerms map[string][]bakery.Op
+
+	// Caveats is a list of caveats that will be added to the default
+	// macaroons.
+	Caveats []checkers.Caveat
+
+	// DBPassword is the password that will be used to encrypt the macaroon
+	// db. If DBPassword is not set, then LndClient, EphemeralKey and
+	// KeyLocator must be set instead.
+	DBPassword []byte
+
+	// LndClient is an LND client that can be used for any lnd queries.
+	// This only needs to be set if DBPassword is not set.
+	LndClient *LndServices
+
+	// EphemeralKey is a key that will be used to derive a shared secret
+	// with LND. This only needs to be set if DBPassword is not set.
+	EphemeralKey *btcec.PublicKey
+
+	// KeyLocator is the locator used to derive a shared secret with LND.
+	// This only needs to be set if DBPassword is not set.
+	KeyLocator *keychain.KeyLocator
+}
+
+func NewMacaroonService(cfg *MacaroonServiceConfig) (*MacaroonService, error) {
+	// Validate config.
+	if cfg.DBPath == "" {
+		return nil, errors.New("no macaroon db path provided")
+	}
+
+	if cfg.MacaroonLocation == "" {
+		return nil, errors.New("no macaroon location provided")
+	}
+
+	if cfg.DBFileName == "" {
+		cfg.DBFileName = defaultDBName
```

Comment:
I'm _somewhat_ more of a fan of having a `DefaultConfig` function which sets all of these values, and then having validation just error out if the values we want aren't set, rather than mutating the config passed in. Not blocking.

Reply:
do you mean: generate default config, then checking each of the passed in config values to see if the default config values should be overwritten and then returning the modified default config struct? feels like that is more bloated cause would then need to check every single config item to see if it has changed from the default. wdyt?

Reply:
I think a default config works well when using CLI flags, as the user is explicitly overwriting values. But since this is a library function, using a default struct does make it more bloated. So I think this is fine the way it is.